modules/reporting/app/workers/cost_query/pdf/timesheet_generator.rb

```diff
@@ -0,0 +1,385 @@
+    @total_page_nr = nil
+    @page_count = 1
+    setup_page!
+  end
+
+  def heading
+    query.name || I18n.t(:"export.timesheet.timesheet")
+  end
+
+  def footer_title
+    heading
+  end
+
+  def cover_page_title
+    "OpenProject"
+  end
+
+  def cover_page_heading
+    heading
+  end
+
+  def cover_page_dates
+    dates_range = all_entries.group_by(&:spent_on).sort
+    start_date = dates_range.first&.first
+    end_date = dates_range.last&.first
+    "#{format_date(start_date)} - #{format_date(end_date)}" if start_date && end_date
+  end
+
+  def cover_page_subheading
+    User.current&.name
+  end
+
+  def project
+    @project
+  end
+
+  def query
+    @query
+  end
+
+  def options
+    {}
+  end
+
+  def setup_page!
+    self.pdf = get_pdf
+    configure_page_size!(:portrait)
+    pdf.title = heading
+  end
+
+  def generate!
+    render_doc
+    if wants_total_page_nrs?
+      @total_page_nr = pdf.page_count + @page_count
+      @page_count = 1
+      setup_page! # clear current pdf
+      render_doc
+    end
+    pdf.render
+  rescue StandardError => e
+    Rails.logger.error { "Failed to generate PDF: #{e} #{e.message}}." }
+    error(I18n.t(:error_pdf_failed_to_export, error: e.message))
+  end
+
+  def render_doc
+    write_cover_page! if with_cover?
+    write_heading!
+    write_hr!
+    write_entries!
+    write_headers!
+    write_footers!
+  end
+
+  def write_entries!
+    all_entries
+      .group_by(&:user)
+      .each do |user, result|
+      write_table(user, result)
+    end
+  end
+
+  def all_entries
+    @all_entries ||= query
+                       .each_direct_result
+                       .map(&:itself)
+                       .filter { |r| r.fields["type"] == "TimeEntry" }
+                       .map { |r| TimeEntry.find(r.fields["id"]) }
+  end
+
+  def build_table_rows(entries)
+    rows = [table_header_columns]
+    entries
+      .group_by(&:spent_on)
+      .sort
+      .each do |spent_on, lines|
+      rows.concat(build_table_day_rows(spent_on, lines))
+    end
+    rows
+  end
+
+  def build_table_day_rows(spent_on, entries)
+    day_rows = []
+    entries.each do |entry|
+      day_rows.push(build_table_row(spent_on, entry))
+      if entry.comments.present?
+        day_rows.push(build_table_row_comment(entry))
+      end
+    end
+    day_rows
+  end
+
+  def build_table_row(spent_on, entry)
+    [
+      { content: format_date(spent_on), rowspan: entry.comments.present? ? 2 : 1 },
+      entry.work_package.subject || "",
+      with_times_column? ? format_spent_on_time(entry) : nil,
+      format_hours(entry.hours),
+      entry.activity&.name || ""
+    ].compact
+  end
+
+  def build_table_row_comment(entry)
+    [{
+      content: entry.comments,
+      text_color: COMMENT_FONT_COLOR,
+      font_style: :italic,
+      colspan: table_columns_widths.size
+    }]
+  end
+
+  def table_header_columns
+    [
+      { content: I18n.t(:"activerecord.attributes.time_entry.spent_on"), rowspan: 1 },
```

Comment:
A shortcut to those would be `TimeEntry.human_attribute_name(:spent_on)`. Not needed to change, might just read a bit nicer